missing direnv errors from log

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

I've got a message saying direnv failed that I can press to see Zed.log, but in the log there isn't any mention of direnv.
<img width="260" alt="Screenshot 2024-12-13 at 15 24 42" src="https://github.com/user-attachments/assets/fe963a6b-f592-4ec6-9d34-696b83273861" />


### Environment

Zed: v0.165.4 (Zed)
OS: macOS 15.1.1
Memory: 32 GiB
Architecture: aarch64

### If applicable, add screenshots or screencasts of the incorrect state / behavior

_No response_

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary>

<!-- Click below this line and paste or drag-and-drop your log-->
```

```
<!-- Click above this line and paste or drag-and-drop your log--></details>

> [Zed.log](https://github.com/user-attachments/files/18195960/Zed.log)

The message you are seeing could only be returned from here:

https://github.com/zed-industries/zed/blob/d54662e683a45617fb66d32d4391729320b9aba3/crates/project/src/direnv.rs#L20-L24

If it was `DirenvError::NonZeroExit`, then the log would contain something about `direnv`, as per:
https://github.com/zed-industries/zed/blob/d54662e683a45617fb66d32d4391729320b9aba3/crates/project/src/direnv.rs#L54-L64

As there is no mention of direnv in the logs, this leads me to believe that we are seeing the `FailedRun` variant of the enum. This means this line should be triggered:
https://github.com/zed-industries/zed/blob/d54662e683a45617fb66d32d4391729320b9aba3/crates/project/src/direnv.rs#L49

So we need to look for errors in the logs and the only one matching is 
```
2024-12-19T10:54:04.216954+01:00 [ERROR] Os { code: 2, kind: NotFound, message: "No such file or directory" }
```

Which is super weird, we check for direnv above:
https://github.com/zed-industries/zed/blob/d54662e683a45617fb66d32d4391729320b9aba3/crates/project/src/direnv.rs#L38-L40

So, the `dir` doesn't exist? I don't understand what is going on honestly

@mrnugget any ideas?

--- a/crates/project/src/environment.rs
+++ b/crates/project/src/environment.rs
@@ -14,8 +14,7 @@ use crate::{
 
 pub struct ProjectEnvironment {
     cli_environment: Option<HashMap<String, String>>,
-    get_environment_task: Option<Shared<Task<Option<HashMap<String, String>>>>>,
-    cached_shell_environments: HashMap<WorktreeId, HashMap<String, String>>,
+    environments: HashMap<WorktreeId, Shared<Task<Option<HashMap<String, String>>>>>,
     environment_error_messages: HashMap<WorktreeId, EnvironmentErrorMessage>,
 }
 
@@ -35,27 +34,26 @@ impl ProjectEnvironment {
 
             Self {
                 cli_environment,
-                get_environment_task: None,
-                cached_shell_environments: Default::default(),
+                environments: Default::default(),
                 environment_error_messages: Default::default(),
             }
         })
     }
 
     #[cfg(any(test, feature = "test-support"))]
-    pub(crate) fn set_cached(
+    pub(crate) fn set_environments(
         &mut self,
         shell_environments: &[(WorktreeId, HashMap<String, String>)],
     ) {
-        self.cached_shell_environments = shell_environments
-            .iter()
-            .cloned()
-            .collect::<HashMap<_, _>>();
+        for (worktree, env) in shell_environments.iter().cloned() {
+            self.environments
+                .insert(worktree, Task::ready(Some(env)).shared());
+        }
     }
 
     pub(crate) fn remove_worktree_environment(&mut self, worktree_id: WorktreeId) {
-        self.cached_shell_environments.remove(&worktree_id);
         self.environment_error_messages.remove(&worktree_id);
+        self.environments.remove(&worktree_id);
     }
 
     /// Returns the inherited CLI environment, if this project was opened from the Zed CLI.
@@ -91,43 +89,34 @@ impl ProjectEnvironment {
         worktree_abs_path: Option<Arc<Path>>,
         cx: &ModelContext<Self>,
     ) -> Shared<Task<Option<HashMap<String, String>>>> {
-        if let Some(task) = self.get_environment_task.as_ref() {
+        if let Some(cli_environment) = self.get_cli_environment() {
+            return cx
+                .spawn(|_, _| async move {
+                    let path = cli_environment
+                        .get("PATH")
+                        .map(|path| path.as_str())
+                        .unwrap_or_default();
+                    log::info!(
+                        "using project environment variables from CLI. PATH={:?}",
+                        path
+                    );
+                    Some(cli_environment)
+                })
+                .shared();
+        }
+
+        let Some((worktree_id, worktree_abs_path)) = worktree_id.zip(worktree_abs_path) else {
+            return Task::ready(None).shared();
+        };
+
+        if let Some(task) = self.environments.get(&worktree_id) {
             task.clone()
         } else {
             let task = self
-                .build_environment_task(worktree_id, worktree_abs_path, cx)
+                .get_worktree_env(worktree_id, worktree_abs_path, cx)
                 .shared();
-
-            self.get_environment_task = Some(task.clone());
+            self.environments.insert(worktree_id, task.clone());
             task
-        }
-    }
-
-    fn build_environment_task(
-        &mut self,
-        worktree_id: Option<WorktreeId>,
-        worktree_abs_path: Option<Arc<Path>>,
-        cx: &ModelContext<Self>,
-    ) -> Task<Option<HashMap<String, String>>> {
-        let worktree = worktree_id.zip(worktree_abs_path);
-
-        let cli_environment = self.get_cli_environment();
-        if let Some(environment) = cli_environment {
-            cx.spawn(|_, _| async move {
-                let path = environment
-                    .get("PATH")
-                    .map(|path| path.as_str())
-                    .unwrap_or_default();
-                log::info!(
-                    "using project environment variables from CLI. PATH={:?}",
-                    path
-                );
-                Some(environment)
-            })
-        } else if let Some((worktree_id, worktree_abs_path)) = worktree {
-            self.get_worktree_env(worktree_id, worktree_abs_path, cx)
-        } else {
-            Task::ready(None)
         }
     }
 
@@ -137,50 +126,40 @@ impl ProjectEnvironment {
         worktree_abs_path: Arc<Path>,
         cx: &ModelContext<Self>,
     ) -> Task<Option<HashMap<String, String>>> {
-        let cached_env = self.cached_shell_environments.get(&worktree_id).cloned();
-        if let Some(env) = cached_env {
-            Task::ready(Some(env))
-        } else {
-            let load_direnv = ProjectSettings::get_global(cx).load_direnv.clone();
+        let load_direnv = ProjectSettings::get_global(cx).load_direnv.clone();
 
-            cx.spawn(|this, mut cx| async move {
-                let (mut shell_env, error_message) = cx
-                    .background_executor()
-                    .spawn({
-                        let cwd = worktree_abs_path.clone();
-                        async move { load_shell_environment(&cwd, &load_direnv).await }
-                    })
-                    .await;
+        cx.spawn(|this, mut cx| async move {
+            let (mut shell_env, error_message) = cx
+                .background_executor()
+                .spawn({
+                    let cwd = worktree_abs_path.clone();
+                    async move { load_shell_environment(&cwd, &load_direnv).await }
+                })
+                .await;
 
-                if let Some(shell_env) = shell_env.as_mut() {
-                    let path = shell_env
-                        .get("PATH")
-                        .map(|path| path.as_str())
-                        .unwrap_or_default();
-                    log::info!(
-                        "using project environment variables shell launched in {:?}. PATH={:?}",
-                        worktree_abs_path,
-                        path
-                    );
-                    this.update(&mut cx, |this, _| {
-                        this.cached_shell_environments
-                            .insert(worktree_id, shell_env.clone());
-                    })
-                    .log_err();
+            if let Some(shell_env) = shell_env.as_mut() {
+                let path = shell_env
+                    .get("PATH")
+                    .map(|path| path.as_str())
+                    .unwrap_or_default();
+                log::info!(
+                    "using project environment variables shell launched in {:?}. PATH={:?}",
+                    worktree_abs_path,
+                    path
+                );
 
-                    set_origin_marker(shell_env, EnvironmentOrigin::WorktreeShell);
-                }
+                set_origin_marker(shell_env, EnvironmentOrigin::WorktreeShell);
+            }
 
-                if let Some(error) = error_message {
-                    this.update(&mut cx, |this, _| {
-                        this.environment_error_messages.insert(worktree_id, error);
-                    })
-                    .log_err();
-                }
+            if let Some(error) = error_message {
+                this.update(&mut cx, |this, _| {
+                    this.environment_error_messages.insert(worktree_id, error);
+                })
+                .log_err();
+            }
 
-                shell_env
-            })
-        }
+            shell_env
+        })
     }
 }
 

--- a/crates/project/src/environment.rs
+++ b/crates/project/src/environment.rs
@@ -132,8 +132,10 @@ impl ProjectEnvironment {
             let (mut shell_env, error_message) = cx
                 .background_executor()
                 .spawn({
-                    let cwd = worktree_abs_path.clone();
-                    async move { load_shell_environment(&cwd, &load_direnv).await }
+                    let worktree_abs_path = worktree_abs_path.clone();
+                    async move {
+                        load_worktree_shell_environment(&worktree_abs_path, &load_direnv).await
+                    }
                 })
                 .await;
 
@@ -189,6 +191,42 @@ impl EnvironmentErrorMessage {
     #[allow(dead_code)]
     fn from_str(s: &str) -> Self {
         Self(String::from(s))
+    }
+}
+
+async fn load_worktree_shell_environment(
+    worktree_abs_path: &Path,
+    load_direnv: &DirenvSettings,
+) -> (
+    Option<HashMap<String, String>>,
+    Option<EnvironmentErrorMessage>,
+) {
+    match smol::fs::metadata(worktree_abs_path).await {
+        Ok(meta) => {
+            let dir = if meta.is_dir() {
+                &*worktree_abs_path
+            } else if let Some(parent) = worktree_abs_path.parent() {
+                parent
+            } else {
+                return (
+                    None,
+                    Some(EnvironmentErrorMessage(format!(
+                        "Failed to load shell environment in {}: not a directory",
+                        worktree_abs_path.display()
+                    ))),
+                );
+            };
+
+            load_shell_environment(&dir, load_direnv).await
+        }
+        Err(err) => (
+            None,
+            Some(EnvironmentErrorMessage(format!(
+                "Failed to load shell environment in {}: {}",
+                worktree_abs_path.display(),
+                err
+            ))),
+        ),
     }
 }
 

--- a/crates/project/src/environment.rs
+++ b/crates/project/src/environment.rs
@@ -40,17 +40,6 @@ impl ProjectEnvironment {
         })
     }
 
-    #[cfg(any(test, feature = "test-support"))]
-    pub(crate) fn set_environments(
-        &mut self,
-        shell_environments: &[(WorktreeId, HashMap<String, String>)],
-    ) {
-        for (worktree, env) in shell_environments.iter().cloned() {
-            self.environments
-                .insert(worktree, Task::ready(Some(env)).shared());
-        }
-    }
-
     pub(crate) fn remove_worktree_environment(&mut self, worktree_id: WorktreeId) {
         self.environment_error_messages.remove(&worktree_id);
         self.environments.remove(&worktree_id);
@@ -89,6 +78,10 @@ impl ProjectEnvironment {
         worktree_abs_path: Option<Arc<Path>>,
         cx: &ModelContext<Self>,
     ) -> Shared<Task<Option<HashMap<String, String>>>> {
+        if cfg!(any(test, feature = "test-support")) {
+            return Task::ready(Some(HashMap::default())).shared();
+        }
+
         if let Some(cli_environment) = self.get_cli_environment() {
             return cx
                 .spawn(|_, _| async move {

--- a/crates/project/src/environment.rs
+++ b/crates/project/src/environment.rs
@@ -204,7 +204,7 @@ async fn load_worktree_shell_environment(
     match smol::fs::metadata(worktree_abs_path).await {
         Ok(meta) => {
             let dir = if meta.is_dir() {
-                &*worktree_abs_path
+                worktree_abs_path
             } else if let Some(parent) = worktree_abs_path.parent() {
                 parent
             } else {

--- a/crates/project/src/project.rs
+++ b/crates/project/src/project.rs
@@ -1207,13 +1207,6 @@ impl Project {
                 .await
                 .unwrap();
 
-            project.update(cx, |project, cx| {
-                let tree_id = tree.read(cx).id();
-                project.environment.update(cx, |environment, _| {
-                    environment.set_environments(&[(tree_id, HashMap::default())])
-                });
-            });
-
             tree.update(cx, |tree, _| tree.as_local().unwrap().scan_complete())
                 .await;
         }

--- a/crates/project/src/project.rs
+++ b/crates/project/src/project.rs
@@ -1210,7 +1210,7 @@ impl Project {
             project.update(cx, |project, cx| {
                 let tree_id = tree.read(cx).id();
                 project.environment.update(cx, |environment, _| {
-                    environment.set_cached(&[(tree_id, HashMap::default())])
+                    environment.set_environments(&[(tree_id, HashMap::default())])
                 });
             });
 

